### DISCUSSION TITLE
Next.js API route to upgrade request into web socket (ws package)

### DISCUSSION THREAD

**Discussion by anton-johansson** [vercel/next.js#53780] [Ideas]
### Goals

I want to be able to use a Next.js API route to upgrade a request into a WebSocket.

### Non-Goals

I don't mean to use Next.js to actually handle the WebSocket, just a means to upgrade a request into a WebSocket.

### Background

I wrote a Discord discussion about it, but no luck yet, which is why I'm opening it here:
https://nextjs-forum.com/post/1137462849862631586

I'm currently using a custom server for Next.js: Koa. With Koa, I have added a route, `/api/web-socket` to upgrade the request into a web socket, via the `ws` packge (utilizing the provided `handleUpgrade` function). This works well. However, my end goal is to get rid of Koa and only utilize Next.js' API routes. Therefore, I've replaced my Koa route with a Next.js API route. The file is called `pages/api/web-socket.ts` and it does the exact same thing. Here is the code:

```ts
import type {NextApiHandler} from "next";
import {handleUpgrade} from "~/server/library/web-socket";

const handler: NextApiHandler = async (request, response) => {
    const upgrade = request.headers.upgrade;
    const connection = request.headers.connection;

    if (upgrade?.toLowerCase() !== "websocket" || connection?.toLowerCase() !== "upgrade") {
        response.status(400).end();
        return;
    }

    handleUpgrade(request);
};

export default handler;
```

My own `handleUpgrade` function simply delegates to the `ws` one:
```ts
export const handleUpgrade = (request: IncomingMessage) => {
    const server = global.webSocket.server;
    server.handleUpgrade(
        request,
        request.socket,
        Buffer.alloc(0),
        ws => {
            server.emit("connection", ws, request);
        },
    );
};
```

When I do this, I get `invalid frame header`. This works perfectly fine with Koa, with basically the same code. Anyone have any idea why? Is there some limitation in the Next.js API routes that prevents me to do this?

To get this to work with Koa, I must also tell it to bypass Koa's built in response handling, via `context.respond = false`. Without this, I get `invalid frame header` with Koa too. Maybe that's the problem with Next.js? Is there a similar option for Next.js? I kind of figured that not setting anything on the response would be sufficient. 😬 But apperantly not.

I use Next.js 13.2.4 and `ws` 8.13.0.

### Proposal

I'm not familiar enough with the stuff that happens behind the scenes in Next.js, so I can't really provide a professional proposal here. But I'm **guessing** we could use something similar to Koa's `context.respond = false`.

**Comment by subhamX** [comment]
Did you find any solutions?

**Reply by anton-johansson** [reply]
No, and I haven’t re-tested with a more recent version either. I still desperately want a solution though, so we can kick out Koa in favor of pure Next.js.

Have you found anything?

**Comment by gin337** [comment]
@subhamX I found a solution that worked for me,

Turns out the issue was caused by a mix of node and tsc. While running the deserver via bun, I had no issues; my build system was using tsc for compilation. For some reason, this broke something internally with socket.io.

Solution:
switched from tsc to tsup
Run project via bun in production (main fix; fix might also be possible with deno)
